code/parachain/frame/cosmwasm/src/ibc.rs

```diff
@@ -15,32 +80,654 @@ impl<T: Config> Pallet<T> {
+			.next()
+			.ok_or_else(|| IbcError::implementation_specific("wrong port_id".to_string()))?;
+		ensure!(
+			prefix_address.next().is_none(),
+			IbcError::implementation_specific("wrong port_id".to_string())
+		);
+		Ok(address)
+	}
+
+	fn create(
+		address: T::AccountIdExtended,
+		contract_info: ContractInfoOf<T>,
+	) -> Result<VmPerContract<T>, IbcError> {
+		let gas = Weight::MAX;
+		let vm = {
+			let runtime =
+				<Pallet<T>>::do_create_vm_shared(gas, InitialStorageMutability::ReadWrite);
+
+			VmPerContract { runtime, address, contract_info }
+		};
+		Ok(vm)
+	}
+
+	fn execute<I, M, V>(vm: &mut V, message: M) -> Result<I::Output, IbcError>
+	where
+		// unfortunately, here is reason
+		// 1. Rust fails to decided that VM is V (which it is)
+		// 2. also without unsafe rust cannot do struct which borrows mut 2 of its items (without
+		// callbacks) 3. so in order to reuse calls either need to build entrypoint.rs like wrappers for
+		// each call 4. but here just closed `cosmwasm_vm + pallet-cosmwasm` on functional level (helped
+		// rust type inference)
+		M: serde::Serialize,
+		I: Input + HasInfo,
+		I::Output: serde::de::DeserializeOwned + ReadLimit + DeserializeLimit,
+		V: cosmwasm_vm::vm::VM + ReadWriteMemory + Has<Env> + Has<MessageInfo> + VMBase,
+		<V as VMBase>::Error: std::fmt::Debug,
+		V::Pointer: for<'x> TryFrom<VmOutputOf<'x, V>, Error = VmErrorOf<V>>,
+		for<'x> Unit: TryFrom<VmOutputOf<'x, V>, Error = VmErrorOf<V>>,
+		for<'x> VmInputOf<'x, V>: TryFrom<AllocateInput<V::Pointer>, Error = VmErrorOf<V>>
+			+ TryFrom<DeallocateInput<V::Pointer>, Error = VmErrorOf<V>>
+			+ TryFrom<CosmwasmCallInput<'x, V::Pointer, I>, Error = VmErrorOf<V>>
+			+ TryFrom<CosmwasmCallWithoutInfoInput<'x, V::Pointer, I>, Error = VmErrorOf<V>>,
+		VmErrorOf<V>:
+			From<ReadableMemoryErrorOf<V>> + From<WritableMemoryErrorOf<V>> + From<ExecutorError>,
+	{
+		cosmwasm_call_serialize::<I, V, M>(vm, &message)
+			.map_err(|err| IbcError::implementation_specific(format!("{:?}", err)))
+	}
+
+	fn on_recv_packet_internal(
+		&self,
+		_output: &mut ModuleOutputBuilder,
+		packet: &ibc::core::ics04_channel::packet::Packet,
+		relayer: &pallet_ibc::Signer,
+	) -> Result<Vec<u8>, IbcError> {
+		let address = Self::port_to_address(&packet.destination_port)?;
+		let contract_info = Self::to_ibc_contract(&address)?;
+
+		let message = IbcPacketReceiveMsg {
+			packet: IbcPacket {
+				data: packet.data.clone().into(),
+				src: IbcEndpoint {
+					port_id: packet.source_port.to_string(),
+					channel_id: packet.source_channel.to_string(),
+				},
+				dest: IbcEndpoint {
+					port_id: packet.destination_port.to_string(),
+					channel_id: packet.destination_channel.to_string(),
+				},
+				sequence: packet.sequence.into(),
+				timeout: Err(IbcError::implementation_specific(
+					"https://app.clickup.com/t/39gjzw1".to_string(),
+				))?,
+			},
+			relayer: Addr::unchecked(relayer.to_string()),
+		};
+		let gas = Weight::MAX;
```

Comment:
why/how are we free to set this as max?

Reply:
it is temporal until IBC callback will provide max gas. sure this is not releasable to Picasso/Composable. not part of this PR to tackle this (until IBC callback ready fixed to provide max gas).